src/core/settings/qgssettingsentry.cpp

```diff
@@ -55,37 +55,89 @@ QString QgsSettingsEntryBase::key( const QStringList &dynamicKeyPartList ) const
   }
 
   if ( dynamicKeyPartList.isEmpty() )
   {
     if ( hasDynamicKey() )
-      QgsLogger::warning( QStringLiteral( "Settings '%1' have a dynamic key but the dynamic key part was not provided" ).arg( completeKey ) );
+      QgsDebugMsg( QStringLiteral( "Settings '%1' have a dynamic key but the dynamic key part was not provided" ).arg( completeKey ) );
 
     return completeKey;
   }
   else
   {
     if ( !hasDynamicKey() )
     {
-      QgsLogger::warning( QStringLiteral( "Settings '%1' don't have a dynamic key, the provided dynamic key part will be ignored" ).arg( completeKey ) );
+      QgsDebugMsg( QStringLiteral( "Settings '%1' don't have a dynamic key, the provided dynamic key part will be ignored" ).arg( completeKey ) );
       return completeKey;
     }
 
     for ( int i = 0; i < dynamicKeyPartList.size(); i++ )
     {
-      completeKey.replace( QString( "%%1" ).arg( QString::number( i + 1 ) ), dynamicKeyPartList.at( i ) );
+      completeKey.replace( QStringLiteral( "%%1" ).arg( QString::number( i + 1 ) ), dynamicKeyPartList.at( i ) );
     }
   }
   return completeKey;
 }
 
+bool QgsSettingsEntryBase::keyIsValid( const QString &key ) const
+{
+  // Key to check
+  QString completeKeyToCheck = key;
+
+  QString settingsPrefix = QgsSettings().prefixedKey( QString(), section() );
+  settingsPrefix.chop( 1 );
+  if ( !completeKeyToCheck.startsWith( settingsPrefix ) )
+  {
+    if ( !mPluginName.isEmpty()
+         && !completeKeyToCheck.startsWith( mPluginName ) )
+    {
+      if ( !completeKeyToCheck.startsWith( '/' ) )
+        completeKeyToCheck.prepend( '/' );
+      completeKeyToCheck.prepend( mPluginName );
+    }
+
+    if ( !completeKeyToCheck.startsWith( '/' ) )
+      completeKeyToCheck.prepend( '/' );
+    completeKeyToCheck.prepend( settingsPrefix );
+  }
+
+  // Prefixed settings key
+  QString prefixedSettingsKey = definitionKey();
+  if ( !prefixedSettingsKey.startsWith( settingsPrefix ) )
+  {
+    if ( !prefixedSettingsKey.startsWith( '/' ) )
+      prefixedSettingsKey.prepend( '/' );
+    prefixedSettingsKey.prepend( settingsPrefix );
+  }
+
+  if ( !hasDynamicKey() )
+    return completeKeyToCheck == prefixedSettingsKey;
+
+  QRegularExpression regularExpression( prefixedSettingsKey.replace( QRegularExpression( QStringLiteral( "%\\d+" ) ), QStringLiteral( ".*" ) ) );
+  QRegularExpressionMatch regularExpresisonMatch = regularExpression.match( completeKeyToCheck );
+  return regularExpresisonMatch.hasMatch();
+}
+
+QString QgsSettingsEntryBase::definitionKey() const
+{
+  QString completeKey = mKey;
+  if ( !mPluginName.isEmpty() )
+  {
+    if ( !completeKey.startsWith( '/' ) )
+      completeKey.prepend( '/' );
+    completeKey.prepend( mPluginName );
+  }
+
+  return completeKey;
+}
+
 bool QgsSettingsEntryBase::hasDynamicKey() const
 {
-  static const QRegularExpression regularExpression( "%\\d+" );
+  static const QRegularExpression regularExpression( QStringLiteral( "%\\d+" ) );
```

Comment:
Apparently QRegularExpression is not thread safe, if there is a chance this can be called by multiple threads you need to use thread_local.